src/index/silentpaymentindex.cpp

```diff
@@ -0,0 +1,135 @@
+#include <index/silentpaymentindex.h>
+
+#include <chainparams.h>
+#include <coins.h>
+#include <common/args.h>
+#include <index/disktxpos.h>
+#include <node/blockstorage.h>
+#include <pubkey.h>
+#include <primitives/transaction.h>
+// TODO: move GetSilentPaymentsTweakDataFromTxInputs out of wallet?
+#include <wallet/silentpayments.h>
+#include <undo.h>
+#include <validation.h>
+
+#include <dbwrapper.h>
+#include <hash.h>
+
+constexpr uint8_t DB_SILENT_PAYMENT_INDEX{'s'};
+
+std::unique_ptr<SilentPaymentIndex> g_silent_payment_index;
+
+/** Access to the silent payment index database (indexes/silentpaymentindex/) */
+class SilentPaymentIndex::DB : public BaseIndex::DB
+{
+public:
+    explicit DB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    bool WriteSilentPayments(const std::vector<std::pair<uint256, CPubKey>>& items);
+};
+
+SilentPaymentIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(gArgs.GetDataDirNet() / "indexes" / "silentpaymentindex", n_cache_size, f_memory, f_wipe)
+{}
+
+bool SilentPaymentIndex::DB::WriteSilentPayments(const std::vector<std::pair<uint256, CPubKey>>& items)
+{
+    CDBBatch batch(*this);
+    for (const auto& tuple : items) {
+        batch.Write(std::make_pair(DB_SILENT_PAYMENT_INDEX, tuple.first), tuple.second);
+    }
+    return WriteBatch(batch);
+}
+
+SilentPaymentIndex::SilentPaymentIndex(std::unique_ptr<interfaces::Chain> chain, size_t n_cache_size, bool f_memory, bool f_wipe)
+    : BaseIndex(std::move(chain), "silentpaymentindex"), m_db(std::make_unique<SilentPaymentIndex::DB>(n_cache_size, f_memory, f_wipe))
+{}
+
+SilentPaymentIndex::~SilentPaymentIndex() {}
+
+// TODO: it's more efficient to do this per block
+bool SilentPaymentIndex::GetSilentPaymentKey(const CTransactionRef& tx, const CBlockIndex& blockindex, CPubKey& tweakedPubKeySum)
+{
+    if (tx->IsCoinBase()) {
+        return false;
+    }
+
+    for (auto& vout : tx->vout) {
+        std::vector<std::vector<unsigned char>> solutions;
+        TxoutType whichType = Solver(vout.scriptPubKey, solutions);
+
+        if (whichType != TxoutType::WITNESS_V1_TAPROOT) {
+            return false;
+        }
+    }
```

Comment:
I think this logic doesn't match the BIP, as it currently implements "all outputs have to be taproot outputs" rather than "at least _one_ output has to be a taproot output"? Could use `std::none_of` to achieve the latter (https://en.cppreference.com/w/cpp/algorithm/all_any_none_of).

Something like
```
    if (std::none_of(tx.vout.begin(), tx.vout.end(), [](const TxOut& txout) {
            std::vector<std::vector<unsigned char>> solutions;
            return Solver(vout.scriptPubKey, solutions) == TxoutType::WITNESS_V1_TAPROOT;
        }) {
        return false;
    }
```
(untested)

Reply:
@josibake it sounds like this should be another helper function. The index shouldn't be too far in the weeds of the BIP rules.

However, I'll look into fixing this.

Reply:
Should be fixed now.

Reply:
Oops, pushing new fix.